libraries/AP_Torqeedo/AP_Torqeedo.cpp

```diff
@@ -0,0 +1,310 @@
+}
+
+// set DE Serial CTS pin to enable sending commands to motor
+void AP_Torqeedo::send_start()
+{
+    // set gpio pin or serial port's CTS pin
+    if (_pin_de > -1) {
+        hal.gpio->write(_pin_de, 1);
+    } else {
+        _uart->set_CTS_pin(true);
+    }
+}
+
+// check for timeout after sending and unset pin if required
+void AP_Torqeedo::check_for_send_end()
+{
+    if (_send_delay_us == 0) {
+        // not sending
+        return;
+    }
+
+    if (AP_HAL::micros() - _last_send_motor_us < _send_delay_us) {
+        // return if delay has not yet elapsed
+        return;
+    }
+    _send_delay_us = 0;
+
+    // unset gpio or serial port's CTS pin
+    if (_pin_de > -1) {
+        hal.gpio->write(_pin_de, 0);
+    } else {
+        _uart->set_CTS_pin(false);
+    }
+}
+
+// calculate delay require to allow bytes to be sent
+uint32_t AP_Torqeedo::calc_send_delay_us(uint8_t num_bytes)
+{
+    // baud rate of 19200 bits/sec
+    // total number of bits = 10 x num_bytes
+    // convert from seconds to micros by multiplying by 1,000,000
+    // plus additional 300us safety margin
+    const uint32_t delay_us = 1e6 * num_bytes * 10 / TORQEEDO_SERIAL_BAUD + 300;
+    return delay_us;
+}
+
+// format of motor speed command packet
+//
+// Data Byte    Field Definition    Example Value   Comments
+// ---------------------------------------------------------------------------------
+// byte 0       Header              0xAC
+// byte 1       SourceId?           0x00            0 = tiller?
+// byte 2       Destination ID?     0x00            0 = all?
+// byte 3       Command Id?         0x05            0=Stop? 4=Don'tTurn? 5=Turn?
+// byte 4       Command Id?         0x00            0x20 if byte3=4, 0x0 is byte3=5
+// byte 5       Motor Speed MSB     ----            Motor Speed MSB (-1000 to +1000)
+// byte 6       Motor Speed LSB     ----            Motor Speed LSB (-1000 to +1000)
+// byte 7       CRC-Maxim           ----            CRC-Maxim value
+// byte 8       Footer              0xAD
+//
+// example message when rotating forwards:  "AC 00 00 05 00 00 ED 95 AD"    (+237)
+// example message when rotating backwards: "AC 00 00 05 00 FF AE 2C 0C AD" (-82)
+
+
+// send a motor speed command as a value from -1000 to +1000
+// value is taken directly from SRV_Channel
+void AP_Torqeedo::send_motor_speed_cmd()
+{
+    // calculate desired motor speed
+    if (!hal.util->get_soft_armed()) {
+        _motor_speed = 0;
+    } else {
+        // convert throttle output to motor output in range -1000 to +1000
+        // ToDo: convert PWM output to motor output so that SERVOx_MIN, MAX and TRIM take effect
+        _motor_speed = constrain_int16(SRV_Channels::get_output_norm(SRV_Channel::Aux_servo_function_t::k_throttle) * 1000.0, -1000, 1000);
+    }
+
+    // set send pin
+    send_start();
+
+    uint8_t mot_speed_cmd_buff[] = {TORQEEDO_PACKET_HEADER, 0x0, 0x0, 0x5, 0x0, HIGHBYTE(_motor_speed), LOWBYTE(_motor_speed), 0x0, TORQEEDO_PACKET_FOOTER};
```

Comment:
Would've been nicer to use a union of a structure and a byte buffer.  This is workable.

Reply:
yes, I just don't know enough about the packets yet to come up with a structure.